tool/net/redbean.c

```diff
@@ -4697,6 +4697,162 @@ static int LuaGetHttpReason(lua_State *L) {
   return 1;
 }
 
+static int EncodeJsonData(lua_State *L, char **buf, int level, char *numformat) {
+  size_t idx = -1;
+  size_t tbllen, buflen;
+  bool isarray;
+  int t = lua_type(L, idx);
+  if (level < 0) return luaL_argerror(L, 1, "too many nested tables");
+  if (LUA_TSTRING == t) {
+    appendf(buf, "\"%s\"", gc(EscapeJsStringLiteral(lua_tostring(L, idx), -1, 0)));
+  } else if (LUA_TNUMBER == t) {
+    // TODO: what if int64_t isn't representable as double
+    appendf(buf, numformat, lua_tonumber(L, idx));
+  } else if (LUA_TBOOLEAN == t) {
+    appends(buf, lua_toboolean(L, idx) ? "true" : "false");
+  } else if (LUA_TTABLE == t) {
+    tbllen = lua_rawlen(L, idx);
+    // encode tables with numeric indices and empty tables as arrays
+    isarray = tbllen > 0 || // integer keys present
+      (lua_pushnil(L), lua_next(L, -2) == 0) || // no non-integer keys
```

Comment:
Lua will clean up the stack on function return if this conditional expression short-circuits, correct?

Reply:
Right; if the second line returns true, then there is nothing left on the stack and if it continues to the third line, then the stack is corrected with lua_pop there.

Reply:
The right thing to do is probably to do nothing. Quoth ANSI X3.159-1988 § 3.2.1.4

```
   When a double is demoted to float or a long double to double or
float, if the value being converted is outside the range of values
that can be represented, the behavior is undefined. 
```

If I write a program like this:

```c
int main(int argc, char *argv[]) {
  volatile int64_t x = INT64_MAX;
  printf("%ld\n%.19g\n", INT64_MAX, (double)x);
  return 0;
}
```

It appears to output an approximation:

```
9223372036854775807
9223372036854775808
```

Reply:
Oh whoops wrong comment.